wasmtime/_memory.py

```diff
@@ -62,6 +64,57 @@ def data_ptr(self, store: Storelike) -> "ctypes._Pointer[c_ubyte]":
         """
         return ffi.wasmtime_memory_data(store._context, byref(self._memory))
 
+    def read(
+            self,
+            store: Storelike,
+            start: typing.Optional[int] = 0,
+            stop: typing.Optional[int] = None) -> typing.Any:
+        """
+        provide fast way to read large memory slice similar to list[start:stop]
+        negative start, stop is allowed in a way similat to list slice mylist[-10:]
+        """
+        data_ptr = self.data_ptr(store)
+        size = self.data_len(store)
+        key = slice(start, stop, None)
+        start, stop, _ = key.indices(size)
+        val_size = stop - start
+        if val_size <= 0:
+            # return bytearray of size zero
+            return bytearray(0)
+        ptr_type = ctypes.c_ubyte * val_size
+        src_ptr = (ptr_type).from_address(ctypes.addressof(data_ptr.contents) + start)
+        return bytearray(src_ptr)
+
+    def write(
+            self,
+            store: Storelike,
+            value: typing.Any,
+            start: typing.Optional[int] = 0,
+            stop: typing.Optional[int] = None) -> typing.Any:
+        """
+        write into a possibly large slice of memory
+        negative start, stop is allowed in a way similat to list slice mylist[-10:]
+        """
+        data_ptr = self.data_ptr(store)
+        size = self.data_len(store)
+        key = slice(start, stop)
+        start, stop, _ = key.indices(size)
+        # value must be bytearray ex. cast bytes() to bytearray
+        if not isinstance(value, array.array) and not isinstance(value, bytearray):
+            # value = array.array('B', value)
+            value = bytearray(value)
+        val_size = len(value)
+        # key.indices(size) knows about size but not val_size
+        stop = start + min(stop - start, val_size)
+        # NOTE: we can use * 1, because we need pointer to the start only
+        ptr_type = ctypes.c_ubyte * val_size
+        src_ptr = (ptr_type).from_buffer(value)
+        dst_ptr = ctypes.addressof(
+            (ptr_type).from_address(ctypes.addressof(data_ptr.contents) + start)
+        )
```

Comment:
This does an addressof of a from_address expression, but is that necessary?

Reply:
correct